### PR TITLE
 nautilus: rgw: GET/HEAD and PUT operations on buckets w/lifecycle expiration configured do not return x-amz-expiration header

### PR DESCRIPTION
**Pull request by pdvian**
https://tracker.ceph.com/issues/41122

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

**Comment by smithfarm** [comment]
Build failure:

```
-- Could not find fmt, will build it
CMake Error at src/CMakeLists.txt:287 (add_subdirectory):
  add_subdirectory given source "fmt" which is not an existing directory.
```

**Comment by tchaikov** [comment]
retest this please.

**Comment by smithfarm** [comment]
Well, to start with, there's no `src/fmt` directory in nautilus...

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -281,12 +281,13 @@ add_subdirectory(json_spirit)
 include_directories(SYSTEM "${CMAKE_SOURCE_DIR}/src/xxHash")
 include_directories(SYSTEM "${CMAKE_SOURCE_DIR}/src/rapidjson/include")
 
+find_package(fmt 5.2.1 QUIET)
+if(NOT fmt_FOUND)
+  message(STATUS "Could not find fmt, will build it")
+  add_subdirectory(fmt)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: we need to continue using `seastar/fmt`.

**Comment by tchaikov** [comment]
@smithfarm thanks for reminder! i just recalled that we lifted `fmt` from `seastar/fmt` after nautilus was released..

File: src/CMakeLists.txt (hunk 1 above)
Comment (pdvian) [reply src/CMakeLists.txt]: @tchaikov Shall I modify this PR to reflect "seastar/fmt" dir ?

Comment (tchaikov) [reply src/CMakeLists.txt]: @pdvian yeah, to be specific. we just need to move the
```cmake
  find_package(fmt 5.2.1 QUIET)		
   if(NOT fmt_FOUND)		
     message(STATUS "Could not find fmt, will build it")		
     add_subdirectory(seastar/fmt)		
   endif()
```
block out of `if(WITH_SEASTAR)`.

**Comment by smithfarm** [comment]
@pdvian Ping, could you please make the change suggested by Kefu in his comment https://github.com/ceph/ceph/pull/30020#discussion_r320545107 ?

File: src/rgw/rgw_lc.cc [hunk 2]
```diff
@@ -19,6 +20,12 @@
 #include "rgw_zone.h"
 #include "rgw_string.h"
 
+// this seems safe to use, at least for now--arguably, we should
+// prefer header-only fmt, in general
+#undef FMT_HEADER_ONLY
+#define FMT_HEADER_ONLY 1
+#include "fmt/format.h"
```
Comment (smithfarm) [inline src/rgw/rgw_lc.cc]: there is no `fmt` in nautilus - closing

**Comment by smithfarm** [comment]
there is no `fmt` submodule in nautilus

**Comment by mattbenjamin** [comment]
in our downstream I experimented with using seastar/fmt as Kefu mentions, had trouble with jenkins, and just put the minimal header-only fmt files in src/rgw/fmt.   n.b., this is considered a valuable feature by applications that use lifecycle expiration

Matt

**Comment by yuvalif** [comment]
pleas note that this PR is required so that the work in this PR: https://github.com/ceph/ceph/pull/31878
could be backported to nautilus.

**Comment by yuvalif** [comment]
this PR: https://github.com/ceph/ceph/pull/32924 was created for backporting the above commits

**Comment by smithfarm** [comment]
closing in favor of #32924

**Pull request closed without merging**